src/services/browser/BrowserSession.ts

```diff
@@ -1,36 +1,95 @@
 import { BrowserSettings } from "../../shared/BrowserSettings"
-// import * as chromeLauncher from "chrome-launcher"
+import { discoverChromeInstances, testBrowserConnection, isPortOpen } from "./BrowserDiscovery"
+import * as chromeLauncher from "chrome-launcher"
+import { Controller } from "../../core/controller"
+import { telemetryService } from "../../services/telemetry/TelemetryService"
 
 interface PCRStats {
 	puppeteer: { launch: typeof launch }
 	executablePath: string
 }
 
-// const DEBUG_PORT = 9222 // Chrome's default debugging port
+// Define browser connection info interface
+export interface BrowserConnectionInfo {
+	isConnected: boolean
+	isRemote: boolean
+	host?: string
+}
+
+const DEBUG_PORT = 9222 // Chrome's default debugging port
 
 export class BrowserSession {
 	private context: vscode.ExtensionContext
 	private browser?: Browser
 	private page?: Page
 	private currentMousePosition?: string
+	private cachedWebSocketEndpoint?: string
+	private lastConnectionAttempt: number = 0
 	browserSettings: BrowserSettings
+	private isConnectedToRemote: boolean = false
+
+	// Telemetry tracking properties
+	private sessionStartTime: number = 0
+	private browserActions: string[] = []
+	private taskId?: string
 
 	constructor(context: vscode.ExtensionContext, browserSettings: BrowserSettings) {
 		this.context = context
 		this.browserSettings = browserSettings
 	}
 
-	private async ensureChromiumExists(): Promise<PCRStats> {
+	// Tests remote browser connection
+	async testConnection(host: string): Promise<{ success: boolean; message: string; endpoint?: string }> {
+		return testBrowserConnection(host)
+	}
+
+	/**
+	 * Get current browser connection information
+	 */
+	getConnectionInfo(): BrowserConnectionInfo {
+		return {
+			isConnected: !!this.browser,
+			isRemote: this.isConnectedToRemote,
+			host: this.isConnectedToRemote ? this.browserSettings.remoteBrowserHost : undefined,
+		}
+	}
+
+	async getDetectedChromePath(): Promise<{ path: string; isBundled: boolean }> {
+		// First check VSCode config
+		const configPath = vscode.workspace.getConfiguration("cline").get<string>("chromeExecutablePath")
+		if (configPath && (await fileExistsAtPath(configPath))) {
+			return { path: configPath, isBundled: false }
+		}
+
+		// Then try to find system Chrome
+		try {
+			const systemPath = chromeLauncher.Launcher.getFirstInstallation()
+			// Add validation to ensure path is not in Trash - This can happen on Mac OS due to the way the chrome-launcher library works
+			if (systemPath && !systemPath.includes(".Trash") && (await fileExistsAtPath(systemPath))) {
+				return { path: systemPath, isBundled: false }
+			}
+		} catch (error) {
+			console.info("Could not find system Chrome:", error)
+		}
+
+		// Finally fall back to PCR's bundled version
+		const stats = await this.ensureChromiumExists()
+		return { path: stats.executablePath, isBundled: true }
+	}
+
+	async ensureChromiumExists(): Promise<PCRStats> {
```

Comment:
This file is very large would it be possible to rip out some code from here to make separate functions 

`ensureChromiumExists` and `getDetectedChromePath` can be ripped out into a separate file